src/main/scala/cromwell/binding/Task.scala

```diff
@@ -52,27 +32,121 @@ object Task {
+   * @return Seq[TaskInput] where TaskInput contains the input
+   *         name & type as well as any postfix quantifiers (?, +)
+   */
+  val inputs: Seq[TaskInput] = for (declaration <- declarations; input <- declaration.asTaskInput) yield input
+  // TODO: I think TaskInput can be replaced by Declaration
+
+  /**
+   * Given a map of task-local parameter names and WdlValues, create a command String.
+   *
+   * Instantiating a command line is the process of taking a command in this form:
+   *
+   * {{{
+   *   sh script.sh ${var1} -o ${var2}
+   * }}}
+   *
+   * This command is stored as a `Seq[CommandPart]` in the `Command` class (e.g. [sh script.sh, ${var1}, -o, ${var2}]).
+   * Then, given a map of variable -> value:
+   *
+   * {{{
+   * {
+   *   "var1": "foo",
+   *   "var2": "bar"
+   * }
+   * }}}
+   *
+   * It calls instantiate() on each part, and passes this map. The ParameterCommandPart are the ${var1} and ${var2}
+   * pieces and they lookup var1 and var2 in that map.
+   *
+   * The command that's returned from Command.instantiate() is:
+   *
+   *
+   * {{{sh script.sh foo -o bar}}}
+   *
+   * @param parameters Parameter values
+   * @return String instantiation of the command
+   */
+  def instantiateCommand(parameters: CallInputs, functions: WdlFunctions = new NoFunctions): Try[String] = {
+    Try(normalize(commandTemplate.map(_.instantiate(declarations, parameters, functions)).mkString("")))
+  }
+
+  def commandTemplateString: String = normalize(commandTemplate.map(_.toString).mkString)
+
   /**
-   * Inputs to this task, as task-local names (i.e. not fully-qualified)
+   * 1) Remove all leading newline chars
+   * 2) Remove all trailing newline AND whitespace chars
+   * 3) Remove all *leading* whitespace that's common among every line in the input string
+   *
+   * For example, the input string:
+   *
+   * "
+   *   first line
+   *     second line
+   *   third line
    *
-   * @return Map of input name to type for that input
+   * "
+   *
+   * Would be normalized to:
+   *
+   * "first line
+   *   second line
+   * third line"
+   *
+   * @param s String to process
+   * @return String which has common leading whitespace removed from each line
    */
-  val inputs: Seq[TaskInput] = {
-    val commandInputs = command.inputs
-    val declarationInputs = for(declaration <- declarations; input <- declaration.asTaskInput) yield input
-    commandInputs ++ declarationInputs
+  private def normalize(s: String): String = {
+    val trimmed = stripAll(s, "\r\n", "\r\n \t")
+    val parts = trimmed.split("[\\r\\n]+")
+    val indent = parts.map(leadingWhitespaceCount).min
+    parts.map(_.substring(indent)).mkString("\n")
   }
 
-  val runtimeAttributes = RuntimeAttributes(ast, backendType)
+  private def leadingWhitespaceCount(s: String): Int = {
+    val matcher = Ws.matcher(s)
+    if (matcher.lookingAt) matcher.end else 0
+  }
+
+  private def stripAll(s: String, startChars: String, endChars: String): String = {
```

Comment:
This also seems like a good candidate for that implicit value class I mentioned above